src/crimson/osd/main.cc

```diff
@@ -38,54 +38,55 @@ seastar::logger& logger() {
 }
 
 void usage(const char* prog) {
-  std::cout << "usage: " << prog << "\n"
-            << "  -i <ID>\n";
+  std::cout << "usage: " << prog << std::endl;
   generic_server_usage();
 }
 
-auto partition_args(seastar::app_template& app, char** argv_begin, char** argv_end)
+auto partition_args(int argc, const char *argv[])
 {
-  // collect all options consumed by seastar::app_template
-  auto parsed = bpo::command_line_parser(std::distance(argv_begin, argv_end),
-                                         argv_begin)
-    .options(app.get_options_description())
-    .style(bpo::command_line_style::default_style &
-           ~bpo::command_line_style::allow_guessing)
-    .allow_unregistered().run();
-  auto unknown_args = bpo::collect_unrecognized(parsed.options,
-                                                bpo::include_positional);
-  std::vector<const char*> ceph_args, app_args;
-  // ceph_argparse_early_args() and
-  // seastar::smp::get_options_description() use "-c" for different
-  // options. and ceph wins
-  auto consume_conf_arg = [&](char** argv) {
-    if (std::strcmp(*argv, "-c") == 0) {
-      std::cout << "warn: apply '-c FILE' as ceph option" << std::endl;
-      ceph_args.push_back(*argv++);
-      if (argv != argv_end) {
-        ceph_args.push_back(*argv++);
-      }
-    }
-    return argv;
-  };
-  auto unknown = unknown_args.begin();
-  auto consume_unknown_arg = [&](char** argv) {
-    for (; unknown != unknown_args.end() &&
-           argv != argv_end &&
-           *unknown == *argv; ++argv, ++unknown) {
-      ceph_args.push_back(*argv);
+  std::vector<const char *> app_args = argv_to_vec(argc, argv);
+  assert(app_args.size() > 0);
+
+  std::vector<const char *> ceph_args;
+  const char* program_name = app_args[0];
+  ConfigValues values;
+
+  class DummyTracker : public ConfigTracker {
+    bool is_tracking(const std::string& name) const override {
```

Comment:
```suggestion
    bool is_tracking(const std::string& name) const final {
```

Reply:
Why mark this as final? Do we really care about a hypothetical derived class overriding is_tracking?

Reply:
A note from the cpp core guidelines:

> [](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#enforcement-125)Note: Use final on functions sparingly. It does not necessarily lead to optimization, and it precludes further overriding.

https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Rh-override

Reply:
@markhpc okay, i am fine either way. i consider `final` a more specific and a more limited form of `override`. and less typing. and if one pronounces it, it might take less time to than "override" does.